File: src/SoulSplitter/Logger.cs
// This file is part of the SoulSplitter distribution (https://github.com/FrankvdStam/SoulSplitter).
// Copyright (c) 2022 Frank van der Stam.
// https://github.com/FrankvdStam/SoulSplitter/blob/main/LICENSE
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

using System;
using System.IO;
using System.Linq;

namespace SoulSplitter
{
    internal static class Logger
    {
        private static bool LoggingEnabled = true;
        private static object _logLock = new object();

        public static void SetLoggingEnabled(bool enable)
        {
            LoggingEnabled = enable;
        }

        public static void Log(string message)
        {
            if(!LoggingEnabled)
            {
                return;
            }

            lock (_logLock)
            {
                try
                {
                    var filePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), @"soulsplitter\soulsplitter.log");
                    var info = new FileInfo(filePath);

                    //Create the file if it does not exist
                    if (!info.Exists)
                    {
                        Directory.CreateDirectory(Path.GetDirectoryName(filePath));
                        File.Create(filePath).Close();
                        info = new FileInfo(filePath);
                    }

                    //If the logfile is bigger than 10 megabytes then delete the first half
                    if (info.Length > 10_485_760)
                    {
                        var lines = File.ReadAllLines(filePath);
                        File.Delete(filePath);
                        File.WriteAllLines(filePath, lines.Skip(lines.Count() / 2));
                    }

                    //Log the message
                    using (var writer = File.AppendText(filePath))
                    {
                        writer.WriteLine($"{DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss:fff")}: {message}");
                        writer.Flush();
                    }
                }
                catch
                {
                    //ignored
                }
            }
        }

        public static void Log(string message, Exception e) => Log(message + " " + e.ToString());
        public static void Log(Exception e) => Log(e.ToString());
    }
}

File: src/soulmods/src/logger.rs
// This file is part of the soulmemory-rs distribution (https://github.com/FrankvdStam/soulmemory-rs).
// Copyright (c) 2022 Frank van der Stam.
// https://github.com/FrankvdStam/soulmemory-rs/blob/main/LICENSE
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

use std::fs;
use std::path::Path;
use log::*;
use log4rs::*;
use log4rs::append::console::ConsoleAppender;
use log4rs::append::file::FileAppender;
use log4rs::encode::pattern::PatternEncoder;
use log4rs::config::{Appender, Config, Logger, Root};

pub fn init_log()
{
    let logfile_path = r#"C:/temp/soulsmods.log"#;

    if Path::new(logfile_path).exists()
    {
        fs::remove_file(logfile_path).unwrap();
    }

    //Setup logger
    let stdout = ConsoleAppender::builder()
        .encoder(Box::new(PatternEncoder::new("{d(%Y-%m-%d %H:%M:%S%.3f)} - {m}{n}")))
        .build();

    let requests = FileAppender::builder()
        .encoder(Box::new(PatternEncoder::new("{d(%Y-%m-%d %H:%M:%S%.3f)} - {m}{n}")))
        .build(logfile_path)
        .unwrap();

    let config = Config::builder()
        .appender(Appender::builder().build("stdout", Box::new(stdout)))
        .appender(Appender::builder().build("log_file", Box::new(requests)))
        .logger(Logger::builder()
            .build("soulsmods", LevelFilter::Info))
        .logger(Logger::builder()
            .appender("log_file")
            .additive(false)
            .build("soulsmods::log_file", LevelFilter::Info))
        .build(Root::builder().appender("stdout").appender("log_file").build(LevelFilter::Info))
        .unwrap();

    let _handle = init_config(config).unwrap();
}
